### PR TITLE
STRINGExplore: Interactive exploration of STRING interactions and associated protein activity

### PR DESCRIPTION
**Pull request by gabmorg**
This tool obtains an interactive network from the input of any gene list, through querying of a (global) high confidence STRING combined score edge list. It uses these edges to find and relate protein action information (also obtained by mapped STRING data) to discovered high-confidence edges in the input system (expected to be a curated gene system and so likely to have high-confidence edges associated with it). Betweenness centrality and tooltips are used to create the network and group the nodes in interactive subsystems.

Future: I'd like to create an option to use coexpression scores from Dr. Steipe's GEO analysis to make the betweenness centrality model more meaningful biologically. I'd also like to create more flexibility in the code for the data sources to allow future iterations of STRING (or even third-party sources) to be compatible with this tool

### CODE REVIEW

File: R/STRINGExplore.R [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  STRINGactions <- fetchData("STRINGactions")
+  STRINGedges <- fetchData("STRINGedges0.8")
+
+  actionSet <- STRINGactions[STRINGactions$protein1 %in% geneSys &
+                               STRINGactions$protein2 %in% geneSys,]
+  sel <- (STRINGedges$protein1 %in% geneSys) & (STRINGedges$protein2 %in% geneSys)
+  geneSysEdges <- STRINGedges[sel, c("protein1", "protein2")]
+
+  #### 2. igraph setup (visNetwork backbone)####
+  sXG <-  graph.data.frame(geneSysEdges, directed = F)
```
Comment (sapirl) [inline R/STRINGExplore.R]: Referring to the package this function belongs to ("igraph::")

File: R/STRINGExplore.R [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  sXGgraph <-  simplify(sXG)
```
Comment (sapirl) [inline R/STRINGExplore.R]: Referring to the package this function belongs to ("igraph::")

File: R/STRINGExplore.R [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+  #### 3. Calculate betweeness centrality of graph based on STRING scores ####
+  bC <-  centr_betw(sXG)
```
Comment (sapirl) [inline R/STRINGExplore.R]: Referring to the package this function belongs to

File: R/STRINGExplore.R [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  nodeBetw <- bC$res
+  nodeBetw <- round(log(nodeBetw + 1)) + 1
+  V(sXGgraph)$btwndegree <- nodeBetw
```
Comment (sapirl) [inline R/STRINGExplore.R]: Referring to the package this function belongs to

File: R/STRINGExplore.R [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+  #### 4. visNetwork setup from igraph object ####
+  # 4.1 Nodes dataframe setup
+  nodes <- get.data.frame(sXGgraph, what="vertices")
```
Comment (sapirl) [inline R/STRINGExplore.R]: Referring to the package this function belongs to

File: R/STRINGExplore.R [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  nodes <- data.frame(id = nodes$name, title = nodes$name,
+                      group = nodes$btwndegree, betweeness = nodes$btwndegree)
+  setNames(nodes, c("id", "title", "betweeness", "betweeness centrality"))
+  nodes <- nodes[order(nodes$id, decreasing = F),]
+
+  # 4.2 Edges dataframe setup
+  edges <- get.data.frame(sXGgraph, what="edges")
```
Comment (sapirl) [inline R/STRINGExplore.R]: Referring to the package this function belongs to

File: R/STRINGExplore.R [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  edges$interactions <- ""
+  for(i in 1:nrow(edges)){
+    edges[i,]$interactions <- getEdgeInteractions(edges[i,]$from, edges[i,]$to, actionSet)
+  }
+  edges$title = paste("<p>",edges$from, "--",edges$to, ":", edges$interactions, "</p>")
+
+  #### 5. Plot visNetwork ####
+  network <- visNetwork(nodes, edges, height = "500px", width = "100%") %>%
```
Comment (sapirl) [inline R/STRINGExplore.R]: Referring to the package this function belongs to

File: dev/mapSTRING.R [hunk 8]
```diff
@@ -0,0 +1,151 @@
+# Script to load STRING coexpression data and map it
```
Comment (sapirl) [inline dev/mapSTRING.R:1]: I can't recall whether Dr. Steipe said that he will upload it or not,
If it is uploaded by him through a different file, don't forget to remove yours!

File: R/STRINGExplore.R (hunk 1 above)
Comment (gabmorg) [reply R/STRINGExplore.R]: used the @import tag approach for the ROxygen header instead

File: R/STRINGExplore.R (hunk 3 above)
Comment (gabmorg) [reply R/STRINGExplore.R]: used the @import tag approach for the ROxygen header instead

**Review by gabmorg — COMMENTED** [review]
Added package::function notation as requested by reviewer, tested STRINGExplore with new code, removed testing code from within fn

**Comment by gabmorg** [comment]
@sapirl addressed changes to code and Travis checks pass. Currently working on vignette and tests, so core code will not be touched. Are you able to merge yet or is that only after the documentations steps?

**Review by gabmorg — COMMENTED** [review]
This is able to automatically merge. Do you mind do that so that I can pull a functional build setup to push my new vignette to?

File: BCB420.2019.ESA.Rproj [hunk 9]
```diff
@@ -18,3 +18,4 @@ StripTrailingWhitespace: Yes
 BuildType: Package
```
Comment (hyginn) [inline BCB420.2019.ESA.Rproj]: `.Rproj` should not be here. Removed.

File: R/STRINGExplore.R [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+#' @export
+
+# it's better to ensure this doesn't rely on internet to run. make the input a gene list,
+# and leverage the exported fn externally (3 inputs: geneSys, STRINGactions, STRINGedges)
+
+STRINGExplore <- function(systemName) {
+  #### 1. Prepare / Load data:####
+  # alternate / future route: add option to create edges from
+  # high correlation scored expression data (Pearson rho >80%) and network those
+  STRINGactions <- fetchData("STRINGactions")
```
Comment (hyginn) [inline R/STRINGExplore.R:93]: Data load should not be done inside functions, and especially not with hardcoded names. Needs to load externally and pass the objects. The closest acceptable approximation to the code as used here would be:
```R
STRINGExplore <- function(systemName, STRINGedges = fetchData("STRINGedges0.8") ) { ...
```

File: R/STRINGExplore.R [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+  STRINGedges <- fetchData("STRINGedges0.8")
+  geneSys <- list()
+
+  dat <- SyDBgetSysSymbols(fetchData("SysDB"), systemName)
```
Comment (hyginn) [inline R/STRINGExplore.R:97]: Same here: needs to be passed from parameter list

File: R/STRINGExplore.R [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+  actionSet <- STRINGactions[STRINGactions$protein1 %in% geneSys &
+                               STRINGactions$protein2 %in% geneSys,]
+  sel <- (STRINGedges$protein1 %in% geneSys) & (STRINGedges$protein2 %in% geneSys)
+  geneSysEdges <- STRINGedges[sel, c("protein1", "protein2")]
+
+  #### 2. igraph setup (visNetwork backbone)####
+  sXG <-  igraph::graph.data.frame(geneSysEdges, directed = F)
+  sXGgraph <-  igraph::simplify(sXG)
+
+  #### 3. Calculate betweeness centrality of graph based on STRING scores ####
```
Comment (hyginn) [inline R/STRINGExplore.R:110]: * betweenness

File: R/STRINGExplore.R [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+  # high correlation scored expression data (Pearson rho >80%) and network those
+  STRINGactions <- fetchData("STRINGactions")
+  STRINGedges <- fetchData("STRINGedges0.8")
+  geneSys <- list()
+
+  dat <- SyDBgetSysSymbols(fetchData("SysDB"), systemName)
+  geneSys <- unlist(dat)
+  names(geneSys) <- c()
+
+  actionSet <- STRINGactions[STRINGactions$protein1 %in% geneSys &
```
Comment (hyginn) [inline R/STRINGExplore.R:101]: better: use a `sel <- ...` idiom here too.

File: dev/STRINGExplore_PHALY-Example.R [hunk 14]
```diff
@@ -0,0 +1,31 @@
+#### Versions ####
```
Comment (hyginn) [inline dev/STRINGExplore_PHALY-Example.R:1]: Two options: move this to `./inst/scripts` after merge, or cover it in the vignette. `./dev/` is only for actual development.

File: dev/mapSTRING.R (hunk 8 above)
Comment (hyginn) [inline dev/mapSTRING.R:1]: This is a part of the documentation. Two options: move this to ./inst/scripts after merge, or cover it in a vignette. ./dev/ is only for actual development.

File: dev/mapSTRING.R [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+# Save the file
+saveRDS(STRINGactions, file = file.path(".", "data", "STRINGactions.RDS"))
+
+# How many of our high-confidence coexpression edges are in our protein actions dataset?
+validationEdges <- apply(STRINGedges[ , c("protein1", "protein2")], 1, sPaste)
+validationActions <- apply(STRINGactions[ , c("protein1", "protein2")], 1, sPaste)
+validate <- intersect(validationActions, validationEdges)
+View(validate)
+length(validate) # 6161 edges with logged protein actions
```
Comment (hyginn) [inline dev/mapSTRING.R:151]: coding style: `# [END]`

**Review by hyginn — COMMENTED** [review]
This is a significant effort, and I am looking forward to see it in action. However I need to cherry pick the files. To avoid issues with other things having gone out of sync, I'll do that manually.

**Comment by hyginn** [comment]
Well - no surprise. Once removing the files that would cause issues after merging, the checks fail. Closing this and merging manually.

**Pull request closed without merging**